File: scripts/node/parsers/globalValueSetParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

function extractAllBlocks(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'g');
  const results = [];
  let m;
  while ((m = re.exec(xml)) !== null) results.push(m[1]);
  return results;
}

/**
 * Reads Global Value Set metadata from SFDX source format.
 * @param {string} typeDir - path to the globalValueSets/ folder
 * @returns {{ valueSets: Array }}
 */
async function parseGlobalValueSets(typeDir) {
  const valueSets = [];
  if (!fs.existsSync(typeDir)) return { valueSets };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.globalValueSet-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    const apiName = file.replace('.globalValueSet-meta.xml', '');

    const valueBlocks = extractAllBlocks(xml, 'customValue');
    const values = valueBlocks.map(block => ({
      value:     extractTag(block, 'fullName'),
      label:     extractTag(block, 'label'),
      isDefault: extractTag(block, 'default') === 'true',
      isActive:  extractTag(block, 'isActive') !== 'false',
    }));

    valueSets.push({
      apiName,
      masterLabel: extractTag(xml, 'masterLabel') || apiName,
      sorted:      extractTag(xml, 'sorted') === 'true',
      values,
    });
  }

  return { valueSets };
}

module.exports = { parseGlobalValueSets };

File: scripts/python/parsers/global_value_set_parser.py
"""Parser for Global Value Set metadata (.globalValueSet-meta.xml)."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def _extract_all_blocks(xml: str, tag: str) -> list[str]:
    return re.findall(rf'<{tag}[^>]*>[\s\S]*?</{tag}>', xml)


def parse_global_value_sets(type_dir: str) -> dict:
    """
    @param type_dir: path to the globalValueSets/ folder
    @returns: { valueSets: list }
    """
    value_sets = []
    p = Path(type_dir)
    if not p.exists():
        return {'valueSets': value_sets}

    for f in sorted(p.glob('*.globalValueSet-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        api_name = f.name.replace('.globalValueSet-meta.xml', '')

        blocks = _extract_all_blocks(xml, 'customValue')
        values = [
            {
                'value':     _extract_tag(b, 'fullName'),
                'label':     _extract_tag(b, 'label'),
                'isDefault': _extract_tag(b, 'default') == 'true',
                'isActive':  _extract_tag(b, 'isActive') != 'false',
            }
            for b in blocks
        ]

        value_sets.append({
            'apiName':     api_name,
            'masterLabel': _extract_tag(xml, 'masterLabel') or api_name,
            'sorted':      _extract_tag(xml, 'sorted') == 'true',
            'values':      values,
        })

    return {'valueSets': value_sets}
